procrastinate/contrib/django/django_connector.py

```diff
@@ -0,0 +1,207 @@
+import asyncio
+import contextlib
+import functools
+import json
+import logging
+import re
+import select
+from json import JSONDecodeError
+from typing import Any, Awaitable, Callable, Iterable, Iterator, Optional
+
+from asgiref.sync import sync_to_async
+from django.db import DatabaseError, IntegrityError
+from django.db import connection as django_connection
+from django.db.backends.postgresql.base import DatabaseWrapper
+from django.utils.connection import ConnectionProxy
+from psycopg2.extensions import connection as pg_conn
+from psycopg2.extras import Json
+
+from procrastinate import connector, exceptions, sql
+from procrastinate.connector import Pool
+
+logger = logging.getLogger(__name__)
+
+LISTEN_TIMEOUT = 30.0
+
+
+def wrap_exceptions(func: Callable) -> Callable:
+    """
+    Wrap django errors as connector exceptions.
+    """
+
+    def _extract_constraint_name(exc: IntegrityError) -> str:
+        msg = str(exc).lower()
+        if "unique constraint" in msg:
+            return msg.split("\n")[0].split("unique constraint")[-1].strip(" \"'")
+        return msg
+
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except IntegrityError as exc:
+            raise exceptions.UniqueViolation(
+                constraint_name=_extract_constraint_name(exc)
+            )
+        except DatabaseError as exc:
+            raise exceptions.ConnectorException from exc
+
+    # Attaching a custom attribute to ease testability and make the
+    # decorator more introspectable
+    wrapped._exceptions_wrapped = True  # type: ignore
+    return wrapped
+
+
+PERCENT_PATTERN = re.compile(r"%(?![\(s])")
+
+
+class DjangoConnector(connector.BaseConnector):
+    @wrap_exceptions
+    def __init__(
+        self,
+        *,
+        json_dumps: Callable = None,
+        json_loads: Callable = None,
+        **kwargs: Any,
+    ):
+        self.json_dumps = json_dumps
+        self.json_loads = json_loads
+
+    def open(self, _=None) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+    async def open_async(self, _=None) -> None:
+        pass
+
+    async def close_async(self) -> None:
+        pass
+
+    def _wrap_json(self, arguments: dict[str, Any]):
+        return {
+            key: Json(value, dumps=self.json_dumps)
+            if isinstance(value, dict)
+            else value
+            for key, value in arguments.items()
+        }
+
+    @contextlib.contextmanager
+    def _connection(self) -> Iterator[ConnectionProxy]:
+        yield django_connection
+
+    def _dict_fetchone(self, cursor) -> dict[str, Any]:
+        """Return row from a cursor as a dict"""
+        columns = [col[0] for col in cursor.description]
+        r = dict(zip(columns, cursor.fetchone()))
+        try:  # TODO fix workaround
+            r["args"] = json.loads(r["args"])
```

Comment:
instead of doing this based on the fact the column is named "args", we could base this on the column OID (`type_code`), but I'm a little surprised that Django hasn't registered the jsonb extension already.

Reply:
It's done [here](https://github.com/django/django/blob/0153a63a674937e4a56d9d5e4ca2d629b011fbde/django/db/backends/postgresql/base.py#L206) in Django

Reply:
I see it's related to this [django 3.1.1 bugfixes](https://docs.djangoproject.com/en/3.1/releases/3.1.1/#bugfixes).

> Fixed a QuerySet.order_by() crash on PostgreSQL when ordering and grouping by JSONField with a custom decoder (#31956). As a consequence, fetching a JSONField with raw SQL now returns a string instead of pre-loaded data. You will need to explicitly call json.loads() in such cases.